File: Segundo/Segundo_Cuatrimestre/Informatica/PDOO/SUBIR/MARCIANITOS_P3/Marcianitos/src/deepspace/SpaceStation.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package deepspace;

import java.util.ArrayList;

/**
 *
 * @author laura
 */
public class SpaceStation {
    static private int MAXFUEL = 100;     
    static private float SHIELDLOSSPERUNITSHOT = 0.1f;
    
    private float ammoPower;
    private float fuelUnits;
    private String name;
    private int nMedals;
    private float shieldPower;
    
    private ArrayList<Weapon> weapons;
    private ArrayList<ShieldBooster> shieldBoosters;
    private Hangar hangar;   
    private Damage pendingDamage;
    
    private void assignFuelValue(float f){
        if (f < MAXFUEL)
            fuelUnits = f;
        else
            fuelUnits = MAXFUEL;
    }
    
    private void cleanPendingDamage(){
        if (getPendingDamage().hasNoEffect())
            pendingDamage = null;        
    }
    
    SpaceStation(String n, SuppliesPackage supplies){
        this.name = n;
        
        // creo q estos se inicializarian a 0 pq si no estamos sumando cosas a algo q no tiene valor
        this.ammoPower = 0;
        this.shieldPower = 0;
        this.fuelUnits = 0;
        
        this.receiveSupplies(supplies);
        
        // estos de abajo creo q se inicializarian así
        this.nMedals = 0; 
        this.weapons = new ArrayList<Weapon> ();
        this.shieldBoosters = new ArrayList<ShieldBooster> ();
        this.hangar = null;
        this.pendingDamage = null; 
    }

	public SpaceStation(SpaceStation otro) {
		this(otro.name, new SuppliesPackage(otro.ammoPower, otro.shieldPower, otro.fuelUnits));
        
        // estos de abajo creo q se inicializarian así
        this.nMedals = otro.nMedals; 
        this.weapons = otro.weapons;
        this.shieldBoosters = otro.shieldBoosters;
        this.hangar = otro.hangar;
        this.pendingDamage = otro.pendingDamage;
		
	}
	
    
    public void cleanUpMountedItems() {
        // Compruebo las armas sin usos y elimino las que no le queden usos
        int num_weapons = this.weapons.size();
        for (int i = 0; i < num_weapons; ++i){
            if (weapons.get(i).getUses() == 0){
                weapons.remove(i);
            }
        }

        //Compruebo los escudos sin usos y elimino los que no tengan usos
        int num_shields = this.shieldBoosters.size();
        for (int i = 0; i < num_shields; ++i){
            if (this.shieldBoosters.get(i).getUses() == 0){
                this.shieldBoosters.remove(i);
            }

        }
    }
    
    public void discardHangar() {
        this.hangar=null;        
    }
    
    public void discardShieldBooster(int i){
        int size = shieldBoosters.size();
        
        if(i>=0 && i<size){
                        
            if(pendingDamage != null){
                
                pendingDamage.discardShieldBooster();
                cleanPendingDamage();
            }
        }    
    }
    
    public void discardShieldBoosterInHangar(int i){
        if (getHangar() != null)
            getHangar().removeShieldBooster(i);
    }
    
    public void discardWeapon(int i){
        
        int size = weapons.size();
        
        if(i>=0 && i<size){
            
            Weapon w = weapons.remove(i);
            
            if(pendingDamage != null){
                
                pendingDamage.discardWeapon(w);
                
                cleanPendingDamage();
            }
        }
    }
    
    public void discardWeaponInHangar(int i){
        if (getHangar() != null)
            getHangar().removeWeapon(i);
    }
    
    public float fire(){
                
        int size = weapons.size();
        
        float factor=1f;
        
        Weapon w;
        
        for(int i=0;i<size;i++){
            
            w = weapons.get(i);
            
            factor *= w.useIt();
                      
        }
        
        return (ammoPower*factor);
    }
    
    public float getAmmoPower(){
        return this.ammoPower;
    }
    
    public float getFuelUnits(){
        return this.fuelUnits;
    }
    
    public Hangar getHangar(){
        return this.hangar;
    }
    
    public String getName(){
        return this.name;
    }
    
    public int getNMedals(){
        return this.nMedals;
    }
    
    public Damage getPendingDamage(){
        return this.pendingDamage;
    }
    
    public ArrayList<ShieldBooster> getShieldBoosters(){
        return this.shieldBoosters;
    }
    
    public float getShieldPower(){
        return this.shieldPower;
    }
    
    public float getSpeed(){
        return (this.fuelUnits/SpaceStation.MAXFUEL);
    }
    
    public SpaceStationToUI getUIversion(){
        return new SpaceStationToUI(this);
    }
    
    public ArrayList<Weapon> getWeapons(){
        return this.weapons;
    }
    
    public void mountShieldBooster(int i){
        ShieldBooster s;
        if (getHangar() != null) {
            s = getHangar().removeShieldBooster(i);
            if (s != null)
                getShieldBoosters().add(s);
        }        
    }
    
    public void mountWeapon(int i){
        Weapon w;
        if (getHangar() != null) {
            w = getHangar().removeWeapon(i);
            if (w != null)
                getWeapons().add(w);
        }
    }
    
    public void move(){
        this.fuelUnits -= this.fuelUnits*this.getSpeed();
    }
    
    public float protection(){
        int size = this.shieldBoosters.size();
        int factor = 1;
        for (int i = 0; i < size; ++i) {
            ShieldBooster s = this.shieldBoosters.get(i);
            factor *= s.useIt();
        }
        return (shieldPower * factor);
    }
    
    public void receiveHangar(Hangar h){
        if (getHangar() == null)    
            this.hangar = h;
    }
    
    public boolean receiveShieldBooster(ShieldBooster s){
        boolean resultado = false;
        
        if (getHangar() != null)  
            resultado = getHangar().addShieldBooster(s);
        
        return resultado; 
    }
    
    public ShotResult receiveShot(float shot){
	ShotResult ret; 
        float myProtection = protection();

        if(myProtection >= shot){
            shieldPower-=SHIELDLOSSPERUNITSHOT*shot;
            shieldPower = Float.max(0f, shieldPower);
            ret = ShotResult.RESIST;
        }
        else{
            shieldPower=0f;
            ret = ShotResult.DONOTRESIST;
        }
        return ret;
    }
    
    public void receiveSupplies(SuppliesPackage s){
        this.ammoPower += s.getAmmoPower();
        this.shieldPower += s.getshieldPower();
        this.assignFuelValue(this.fuelUnits + s.getfuelUnits());
    }
    
    public boolean receiveWeapon(Weapon w){
        boolean resultado = false;
        
        if (getHangar() != null)  
            resultado = getHangar().addWeapon(w);
        
        return resultado;
    }
    
    public Transformation setLoot(Loot loot) {
        
        CardDealer dealer = CardDealer.getInstance();

        // Aniadimos Hangars
        int h = loot.getNHangars();
        if (h > 0){
            hangar = dealer.nextHangar();

            this.receiveHangar(hangar);
        }

        // Aniadimos supplies
        int elements = loot.getNSupplies();
        for (int i = 0; i < elements; ++i){
            SuppliesPackage sup = dealer.nextSuppliesPackage();
            receiveSupplies(sup);
        }

        // Aniadimos Weapons
        elements = loot.getNWeapons();
        for (int i = 0; i < elements; ++i){
            Weapon weap = dealer.nextWeapon();
            receiveWeapon(weap);
        }

        // Aniadimos Shields
        elements = loot.getNShields();
        for (int i = 0; i < elements; ++i){
            ShieldBooster sh = dealer.nextShieldBooster();

            receiveShieldBooster(sh);
        }

        int medals = loot.getNMedals();

        this.nMedals += medals;
		
		Transformation result = Transformation.NOTRANSFORM;
		
		if (loot.getEfficient()){
			result = Transformation.GETEFFICIENT;
		} else if (loot.spaceCity()) {
			result = Transformation.SPACECITY;
		}
		
		return (result);
    }
    
    public void setPendingDamage(Damage d){
        this.pendingDamage = d.adjust(weapons, shieldBoosters);
    }
    
    public boolean validState(){
        return (this.pendingDamage == null || this.pendingDamage.hasNoEffect());
    }
}

File: Segundo/Segundo_Cuatrimestre/Informatica/PDOO/SUBIR/MARCIANITOS_P3/RUBY/src/SpaceStation.rb
#encoding: UTF-8

require_relative 'SpaceStationToUI'

require_relative 'Weapon'
require_relative 'WeaponType'
require_relative 'ShieldBooster'
require_relative 'Hangar'
require_relative 'Damage'
require_relative 'SuppliesPackage'
require_relative 'ShotResult'
require_relative 'Loot'

module Deepspace
  
	class SpaceStation

		#atributos de clase
		@@MAXFUEL=100
		@@SHIELDLOSSPERUNITSHOT=0.1
	  
		attr_reader :name, :nMedals, :ammoPower, :fuelUnits, \
		  :shieldPower, :shieldBoosters, :hangar, :weapons, \
		  :pendingDamage, :MAXFUEL, :SHIELDLOSSPERUNITSHOT
	  
		def initialize (n, supplies)	# n string, supplies es SuppliesPackage	
			@name=n
			@ammoPower = 0
			@shieldPower = 0
			@fuelUnits = 0
			receiveSupplies(supplies)
			
			@nMedals=0
			@weapons=Array.new()
			@shieldBoosters=Array.new()
			@hangar=nil
			@pendingDamage=nil
		end
		
		def cleanUpMountedItems()  # void()
			#Compruebo las armas sin usos y elimino las que no le queden usos
			@weapons.each do |w|
				if (w.uses == 0)
					@weapons.delete(w) #creo q es asi
				end
			end

			#Compruebo los escudos sin usos y elimino los que no tengan usos
			@shieldBoosters.each do |s|
				if (s.uses == 0)
					@shieldBoosters.delete(s) #creo q es asi
				end
			end
		end
		
		def discardHangar()  # void()
			@hangar=nil
		end
		
		def discardShieldBooster(i)   # i es indice del arma
			size = @shieldBoosters.length
			if (i>=0 && i<size)
				s=@shieldBoosters.delete_at(i)
				
				if (@pendingDamage != nil)
					@pendingDamage.discardShieldBooster(w)
					cleanPendingDamage
				end
			end	
		end        #void()
		
		def discardShieldBoosterInHangar(i)  # i es indice del escudo que se descarta
			if (hangar != nil)
				hangar.removeShieldBooster(i) 
			end		# void()
		end
		
		def discardWeapon(i)  # i es indice del arma
		
			size=@weapons.length
			if (i>=0 && i<size)
				w=@weapons.delete_at(i)
				
				if (@pendingDamage != nil)
					@pendingDamage.discardWeapon(w)
					cleanPendingDamage
				end
			end		
		end      # void()
		
		def discardWeaponInHangar(i)  # i es indice del arma que se descarta
			if (@hangar != nil)
				@hangar.removeWeapon(i)
			end         # void()
		end
		
		def fire()
			factor = 1
			
			@weapons.each do |w|
				factor *= w.useIt()
			end
			
			return (ammoPower*factor)  #float
		
		end
		
		def getNMedals()
			return @nMedals
		end
		
		def getSpeed() 
			return (@fuelUnits.to_f/@@MAXFUEL) # devuelve float
		end
		
		def getUIversion()
			SpaceStationToUI.new(self)
		end
		
		def mountShieldBooster(i)  # i es indice dentro del Hangar
			if (hangar != nil) 
				s = hangar.removeShieldBooster(i)
				if (s != nil)
					shieldBoosters.push(s)
				end        # void()
			end
		end
		
		def mountWeapon(i)   # i es indice dentro del Hangar
			if (hangar != nil) 
				w = hangar.removeWeapon(i)
				if (w != nil)
					weapons.push(w)
				end
			end       # void()
		end
		 
		def move()   # void()
			@fuelUnits -= @fuelUnits*getSpeed()
		end
		
		def protection()  #float
			factor = 1
			shieldBoosters.each do |s|
				factor *= s.useIt
			end	
			return (@shieldPower * factor)		
		end
		
		def receiveHangar(h)  # h es un Hangar
			if (hangar == nil)
				@hangar = h
			end   # es void()
		end
		
		def receiveShieldBooster(s)  # s es ShieldBooster
			retorno = false
			
			if (hangar != nil)
				retorno = hangar.addShieldBooster(s)	#creo q asi devuelve true o false ya
			end
			
			return (retorno)   # bool si consigue añadir el escudo
		end
		
		def receiveShot(shot)  # shot es float
			myProtection=protection()
			if(myProtection>=shot)
				@shieldPower=@shieldPower-(@@SHIELDLOSSPERUNITSHOT*shot)
				@shieldPower=[0.0,@shieldPower].max
				ShotResult::RESIST
			else
				@shieldPower=0.0
				ShotResult::DONOTRESIST   # devuelve shotResult
			end
		
		end
		
		def receiveSupplies(s) 
			@ammoPower += s.ammoPower
			@shieldPower += s.shieldPower
			
			assignFuelValue(fuelUnits + s.fuelUnits)
			
		end
		
		def receiveWeapon(w)  # w es Weapon
			retorno = false
			if (hangar != nil)
				retorno = hangar.addWeapon(w)	#creo q asi devuelve true o false ya
			end  # bool si ha conseguido añadir el arma
		end
		
		def setLoot(loot)  #loot es Loot
			dealer = CardDealer.instance
			
			#Aniadimos Hangars
			h = loot.nHangars
			if (h > 0)
				hangar = dealer.nextHangar()
				
				receiveHangar(hangar)  
			end
			
			#Aniadimos supplies
			elements = loot.nSupplies
			for i in (1..elements) do
				sup = dealer.nextSuppliesPackage()
				receiveSupplies(sup)
			end
			
			# Aniadimos Weapons
			elements = loot.nWeapons
			for i in (1..elements) do
				weap = dealer.nextWeapon()
				receiveWeapon(weap)
			end
			
			#Aniadimos Shields
			elements = loot.nShields
			for i in (1..elements) do
				sh = dealer.nextShieldBooster()
				receiveShieldBooster(sh)
			end
			
			medals = loot.nMedals

			@nMedals += medals

		end  # void()
		
		def setPendingDamage(d)   # d es Damage
			#d.adjust(@weapons, @shieldBoosters)
			# se almacena el resultado en el atributo correspondiente ???
			# creo q es esto:
			@pendingDamage = d.adjust(@weapons, @shieldBoosters)      # void()
		end
		
		def validState()
			return (@pendingDamage == nil || @pendingDamage.hasNoEffect())   # bool 
		end
		
		def to_s
			out="Space Station - Name: #{@name}\n"
			out+="\tnMedals: #{@nMedals}, Fuel units: #{@fuelUnits.round(2)}, Power: #{@ammoPower}, Shields: #{@shieldPower}\n"
			out+="\tWeapons: [#{@weapons.join(' // ')}]\n"
			out+="\tShieldBooster: [#{@shieldBoosters.join(' // ')}]\n"
			out+="\tHangars: #{@hangar}\n"
			out+="\tPendingDamage: #{@pendingDamage}\n" 
			out+="------- end of Space Station >> #{@name} << -------"
			return out
		end
		
		private 	#todos los metodos q haya debajo de esto son privados (solo los de instancia)
		def assignFuelValue(f)  # f es float
			if (f < @@MAXFUEL)
				@fuelUnits = f
			else
				@fuelUnits = @@MAXFUEL
			end
		end      # no devulve nada
		
		def cleanPendingDamage()
			if (@pendingDamage.hasNoEffect())
				pendingDamage = nil 
			end 
		end  # void

	end # class
	
	#s = SpaceStation.new("space", SuppliesPackage.new(3,4,5))
	#puts s.to_s
	
	#sui = SpaceStationToUI.new(s)
	#puts sui.to_s
end # module
